lua/luasnip/util/util.lua

```diff
@@ -151,6 +151,51 @@ local function wrap_value(value)
 	return { value }
 end
 
+local LUASNIP_LAST_SELECTION = "LUASNIP_LAST_SELECTION"
+
+local function store_selection()
+	local start_line, start_col = vim.fn.line("'<"), vim.fn.col("'<")
+	local end_line, end_col = vim.fn.line("'>"), vim.fn.col("'>")
+	local mode = vim.fn.visualmode()
+	if
+		not vim.o.selection == "exclusive"
+		and not (start_line == end_line and start_col == end_col)
+	then
+		end_col = end_col - 1
+	end
+
+	local function _vim_line(lnum)
+		return vim.api.nvim_buf_get_lines(0, lnum, lnum + 1, true)[1]
+	end
+	local chunks = {}
+	if start_line == end_line then
+		chunks = { _vim_line(start_line - 1):sub(start_col, end_col + 1) }
+	else
+		local first_col = 0
+		local last_col = nil
+		if mode:lower() ~= "v" then -- mode is block
+			first_col = start_col
+			last_col = end_col
+		end
+
+		chunks = { _vim_line(start_line - 1):sub(start_col, last_col) }
+
+		for cl = start_line, (end_line - 2) do
+			table.insert(chunks, _vim_line(cl):sub(first_col, last_col))
+		end
+		table.insert(
+			chunks,
+			_vim_line(end_line - 1):sub(first_col, end_col + 1)
+		)
+	end
+	vim.api.nvim_buf_set_var(0, LUASNIP_LAST_SELECTION, chunks)
+end
+
+local function get_selection()
+	local ok, val = pcall(vim.api.nvim_buf_get_var, 0, LUASNIP_LAST_SELECTION)
+	return ok and val or ""
+end
+
```

Comment:
Very nice, but wouldn't eg. mapping `<tab>` to `s` and storing the unnamed register do the same?

Reply:
This approach allows having a distinct selection per buffer, I borrowed it mostly from Ultisnips.

Reply:
Per-buffer is great, and that part I wouldn't change, but I think that the function that gets the text could be simplified by deleting the text(maybe not with `s`, `d` would probably be better as we'd stay in normal) and then storing the unnamed register into a buffer-local variable.

Reply:
(I thought ultisnips does it exactly that way tbh)

Reply:
No, ultisnips do almost what I did, the only difference is that it does not use a buffer variable to store the text.
https://github.com/SirVer/ultisnips/blob/8803960a4e09daf1ad4d82e16a25bbdb4c78530e/pythonx/UltiSnips/vim_state.py#L120

https://github.com/SirVer/ultisnips/blob/e96733b5db27b48943db86dd8623f1497b860bc6/autoload/UltiSnips/map_keys.vim#L64

Reply:
Using a buffer vars frees me from having to clean the memorized value myself.

Reply:
Anyway, I will try to replicate the current behavior but using    `s<cmd>...`  and fetching the text from the register, which should be similar to your proposal and will remove the manual vim dance.

Reply:
Thank you! :D
(I do wonder though, why does ultisnips do the 'vim dance'?)

Reply:
Done!, took a bit as I was having lunch.

Reply:
Sweet, thank you!